### ISSUE TITLE
Hook from a template with mutex does not release lock after completion

### ISSUE THREAD

**Issue by mnarozny** [argoproj/argo-workflows#11117]
### Pre-requisites

- [X] I have double-checked my configuration
- [X] I can confirm the issues exists when I tested with `:latest`
- [ ] I'd like to contribute the fix myself (see [contributing guide](https://github.com/argoproj/argo-workflows/blob/master/docs/CONTRIBUTING.md))

### What happened/what you expected to happen?

**In a template using steps, after completing a hook from a template with synchornization mutex, the lock does not get released.**

We have hooks that are making API calls, sometimes the Running hook completes after the Success one is done so we wanted to use the synchronization feature on the hook template to avoid that behaviour. Workflow hangs forever on lock.

Other observations:
- This seems to happen whether the steps are paralel or sequential
- If the mutex is removed, the workflow completes but the behaviour is odd anyway, hooks fire in sequence even though the jobs are starting in paralel
- When using a dag:
   - the mutex works as long as the same mutex name is not used across different steps (locks get stuck as well otherwise)
   - all the 'running' hooks fire at the same time in this scenario but then some of the success hooks are not send out (unless there's a sleep step added at the end, argo-workflows seems to be closing the workflow and ignoring the success hooks)

### Version

v3.4.7

### Paste a small workflow that reproduces the issue. We must be able to run the workflow; don't enter a workflows that uses private images.

```YAML
apiVersion: argoproj.io/v1alpha1
kind: Workflow
metadata:
  generateName: example-steps-simple-mutex
spec:
  entrypoint: main

  templates:
    - name: main
      steps:
        - - name: job-1
            template: sleep
            arguments:
              parameters:
                - name: sleep_duration
                  value: 5
            hooks:
              running:
                expression: steps['job-1'].status == "Running"
                template: hook
                arguments:
                  parameters:
                    - name: sleep_duration
                      value: 5
                    - name: context
                      value: "job-1"
              success:
                expression: steps['job-1'].status == "Succeeded"
                template: hook
                arguments:
                  parameters:
                    - name: sleep_duration
                      value: 5
                    - name: context
                      value: "job-1"
          - name: job-2
            template: sleep
            arguments:
              parameters:
                - name: sleep_duration
                  value: 5
            hooks:
              running:
                expression: steps['job-2'].status == "Running"
                template: hook
                arguments:
                  parameters:
                    - name: sleep_duration
                      value: 5
                    - name: context
                      value: "job-2"
              success:
                expression: steps['job-2'].status == "Succeeded"
                template: hook
                arguments:
                  parameters:
                    - name: sleep_duration
                      value: 5
                    - name: context
                      value: "job-2"
          - name: job-3
            template: sleep
            arguments:
              parameters:
                - name: sleep_duration
                  value: 5
            hooks:
              running:
                expression: steps['job-3'].status == "Running"
                template: hook
                arguments:
                  parameters:
                    - name: sleep_duration
                      value: 45
                    - name: context
                      value: "job-3"
              success:
                expression: steps['job-3'].status == "Succeeded"
                template: hook
                arguments:
                  parameters:
                    - name: sleep_duration
                      value: 5
                    - name: context
                      value: "job-3"

    - name: hook
      synchronization:
        mutex:
          name: "{{ inputs.parameters.context }}"
      inputs:
        parameters:
          - name: sleep_duration
          - name: context
      script:
        image: alpine:latest
        command: [/bin/sh]
        source: |
          echo "Fake API call - sleeping for {{ inputs.parameters.sleep_duration }}, context is {{ inputs.parameters.context }}"
          sleep {{ inputs.parameters.sleep_duration }}
    
    - name: sleep
      inputs:
        parameters:
          - name: sleep_duration
      script:
        image: alpine:latest
        command: [/bin/sh]
        source: |
          echo "Sleeping for {{ inputs.parameters.sleep_duration }}"
          sleep {{ inputs.parameters.sleep_duration }}
```

### Logs from the workflow controller

```text
time="2023-05-22T14:09:50.845Z" level=info msg="Processing workflow" namespace=ci-workflows workflow=example-steps-simple-mutex4b72g
time="2023-05-22T14:09:50.853Z" level=info msg="Updated phase  -> Running" namespace=ci-workflows workflow=example-steps-simple-mutex4b72g
time="2023-05-22T14:09:50.854Z" level=info msg="Steps node example-steps-simple-mutex4b72g initialized Running" namespace=ci-workflows workflow=example-steps-simple-mutex4b72g
time="2023-05-22T14:09:50.854Z" level=info msg="StepGroup node example-steps-simple-mutex4b72g-681339196 initialized Running" namespace=ci-workflows workflow=example-steps-simple-mutex4b72g
time="2023-05-22T14:09:50.854Z" level=info msg="Pod node example-steps-simple-mutex4b72g-531693655 initialized Pending" namespace=ci-workflows workflow=example-steps-simple-mutex4b72g
time="2023-05-22T14:09:50.884Z" level=info msg="Created pod: example-steps-simple-mutex4b72g[0].job-1 (example-steps-simple-mutex4b72g-sleep-531693655)" namespace=ci-workflows workflow=example-steps-simple-mutex4b72g
time="2023-05-22T14:09:50.884Z" level=info msg="Pod node example-steps-simple-mutex4b72g-548471274 initialized Pending" namespace=ci-workflows workflow=example-steps-simple-mutex4b72g
time="2023-05-22T14:09:50.908Z" level=info msg="Created pod: example-steps-simple-mutex4b72g[0].job-2 (example-steps-simple-mutex4b72g-sleep-548471274)" namespace=ci-workflows workflow=example-steps-simple-mutex4b72g
time="2023-05-22T14:09:50.908Z" level=info msg="Pod node example-steps-simple-mutex4b72g-565248893 initialized Pending" namespace=ci-workflows workflow=example-steps-simple-mutex4b72g
time="2023-05-22T14:09:50.933Z" level=info msg="Created pod: example-steps-simple-mutex4b72g[0].job-3 (example-steps-simple-mutex4b72g-sleep-565248893)" namespace=ci-workflows workflow=example-steps-simple-mutex4b72g
time="2023-05-22T14:09:50.933Z" level=info msg="Workflow step group node example-steps-simple-mutex4b72g-681339196 not yet completed" namespace=ci-workflows workflow=example-steps-simple-mutex4b72g
time="2023-05-22T14:09:50.933Z" level=info msg="TaskSet Reconciliation" namespace=ci-workflows workflow=example-steps-simple-mutex4b72g
time="2023-05-22T14:09:50.933Z" level=info msg=reconcileAgentPod namespace=ci-workflows workflow=example-steps-simple-mutex4b72g
time="2023-05-22T14:09:50.950Z" level=info msg="Workflow update successful" namespace=ci-workflows phase=Running resourceVersion=365434517 workflow=example-steps-simple-mutex4b72g
time="2023-05-22T14:10:00.886Z" level=info msg="Processing workflow" namespace=ci-workflows workflow=example-steps-simple-mutex4b72g
time="2023-05-22T14:10:00.886Z" level=info msg="Task-result reconciliation" namespace=ci-workflows numObjs=3 workflow=example-steps-simple-mutex4b72g
time="2023-05-22T14:10:00.886Z" level=info msg="task-result changed" namespace=ci-workflows nodeID=example-steps-simple-mutex4b72g-565248893 workflow=example-steps-simple-mutex4b72g
time="2023-05-22T14:10:00.886Z" level=info msg="task-result changed" namespace=ci-workflows nodeID=example-steps-simple-mutex4b72g-548471274 workflow=example-steps-simple-mutex4b72g
time="2023-05-22T14:10:00.886Z" level=info msg="task-result changed" namespace=ci-workflows nodeID=example-steps-simple-mutex4b72g-531693655 workflow=example-steps-simple-mutex4b72g
time="2023-05-22T14:10:00.886Z" level=info msg="node changed" namespace=ci-workflows new.message= new.phase=Running new.progress=0/1 nodeID=example-steps-simple-mutex4b72g-565248893 old.message= old.phase=Pending old.progress=0/1 workflow=example-steps-simple-mutex4b72g
time="2023-05-22T14:10:00.886Z" level=info msg="node changed" namespace=ci-workflows new.message= new.phase=Running new.progress=0/1 nodeID=example-steps-simple-mutex4b72g-531693655 old.message= old.phase=Pending old.progress=0/1 workflow=example-steps-simple-mutex4b72g
time="2023-05-22T14:10:00.886Z" level=info msg="node changed" namespace=ci-workflows new.message= new.phase=Running new.progress=0/1 nodeID=example-steps-simple-mutex4b72g-548471274 old.message= old.phase=Pending old.progress=0/1 workflow=example-steps-simple-mutex4b72g
time="2023-05-22T14:10:00.887Z" level=info msg="Running hooks" hookName=running lifeCycleHook=running namespace=ci-workflows node="example-steps-simple-mutex4b72g[0].job-1.hooks.running" workflow=example-steps-simple-mutex4b72g
time="2023-05-22T14:10:00.887Z" level=info msg="ci-workflows/Mutex/job-1 acquired by ci-workflows/example-steps-simple-mutex4b72g/example-steps-simple-mutex4b72g-3623933960. Lock availability: 0/1" mutex=ci-workflows/Mutex/job-1
time="2023-05-22T14:10:00.888Z" level=info msg="Node example-steps-simple-mutex4b72g[0].job-1.hooks.running acquired synchronization lock" namespace=ci-workflows workflow=example-steps-simple-mutex4b72g
time="2023-05-22T14:10:00.888Z" level=info msg="Pod node example-steps-simple-mutex4b72g-3623933960 initialized Pending" namespace=ci-workflows workflow=example-steps-simple-mutex4b72g
time="2023-05-22T14:10:00.892Z" level=info msg="cleaning up pod" action=terminateContainers key=ci-workflows/example-steps-simple-mutex4b72g-sleep-565248893/terminateContainers
time="2023-05-22T14:10:00.938Z" level=info msg="Created pod: example-steps-simple-mutex4b72g[0].job-1.hooks.running (example-steps-simple-mutex4b72g-hook-3623933960)" namespace=ci-workflows workflow=example-steps-simple-mutex4b72g
time="2023-05-22T14:10:00.938Z" level=info msg="Workflow step group node example-steps-simple-mutex4b72g-681339196 not yet completed" namespace=ci-workflows workflow=example-steps-simple-mutex4b72g
time="2023-05-22T14:10:00.938Z" level=info msg="TaskSet Reconciliation" namespace=ci-workflows workflow=example-steps-simple-mutex4b72g
time="2023-05-22T14:10:00.938Z" level=info msg=reconcileAgentPod namespace=ci-workflows workflow=example-steps-simple-mutex4b72g
time="2023-05-22T14:10:00.964Z" level=info msg="Workflow update successful" namespace=ci-workflows phase=Running resourceVersion=365435426 workflow=example-steps-simple-mutex4b72g
time="2023-05-22T14:10:10.938Z" level=info msg="Processing workflow" namespace=ci-workflows workflow=example-steps-simple-mutex4b72g
time="2023-05-22T14:10:10.939Z" level=info msg="Task-result reconciliation" namespace=ci-workflows numObjs=4 workflow=example-steps-simple-mutex4b72g
time="2023-05-22T14:10:10.939Z" level=info msg="task-result changed" namespace=ci-workflows nodeID=example-steps-simple-mutex4b72g-3623933960 workflow=example-steps-simple-mutex4b72g
time="2023-05-22T14:10:10.939Z" level=info msg="node changed" namespace=ci-workflows new.message= new.phase=Running new.progress=0/1 nodeID=example-steps-simple-mutex4b72g-3623933960 old.message= old.phase=Pending old.progress=0/1 workflow=example-steps-simple-mutex4b72g
time="2023-05-22T14:10:10.939Z" level=info msg="node changed" namespace=ci-workflows new.message= new.phase=Succeeded new.progress=0/1 nodeID=example-steps-simple-mutex4b72g-548471274 old.message= old.phase=Running old.progress=0/1 workflow=example-steps-simple-mutex4b72g
time="2023-05-22T14:10:10.939Z" level=info msg="node changed" namespace=ci-workflows new.message= new.phase=Succeeded new.progress=0/1 nodeID=example-steps-simple-mutex4b72g-531693655 old.message= old.phase=Running old.progress=0/1 workflow=example-steps-simple-mutex4b72g
time="2023-05-22T14:10:10.939Z" level=info msg="node changed" namespace=ci-workflows new.message= new.phase=Succeeded new.progress=0/1 nodeID=example-steps-simple-mutex4b72g-565248893 old.message= old.phase=Running old.progress=0/1 workflow=example-steps-simple-mutex4b72g
time="2023-05-22T14:10:10.940Z" level=info msg="Running hooks" hookName=success lifeCycleHook=success namespace=ci-workflows node="example-steps-simple-mutex4b72g[0].job-1.hooks.success" workflow=example-steps-simple-mutex4b72g
time="2023-05-22T14:10:10.940Z" level=info msg="Pod node example-steps-simple-mutex4b72g-2845957028 initialized Pending" namespace=ci-workflows workflow=example-steps-simple-mutex4b72g
time="2023-05-22T14:10:10.940Z" level=info msg="Workflow step group node example-steps-simple-mutex4b72g-681339196 not yet completed" namespace=ci-workflows workflow=example-steps-simple-mutex4b72g
time="2023-05-22T14:10:10.940Z" level=info msg="TaskSet Reconciliation" namespace=ci-workflows workflow=example-steps-simple-mutex4b72g
time="2023-05-22T14:10:10.940Z" level=info msg=reconcileAgentPod namespace=ci-workflows workflow=example-steps-simple-mutex4b72g
time="2023-05-22T14:10:10.945Z" level=info msg="cleaning up pod" action=terminateContainers key=ci-workflows/example-steps-simple-mutex4b72g-hook-3623933960/terminateContainers
time="2023-05-22T14:10:10.955Z" level=info msg="Workflow update successful" namespace=ci-workflows phase=Running resourceVersion=365436306 workflow=example-steps-simple-mutex4b72g
time="2023-05-22T14:10:10.961Z" level=info msg="cleaning up pod" action=labelPodCompleted key=ci-workflows/example-steps-simple-mutex4b72g-sleep-531693655/labelPodCompleted
time="2023-05-22T14:10:10.961Z" level=info msg="cleaning up pod" action=labelPodCompleted key=ci-workflows/example-steps-simple-mutex4b72g-sleep-548471274/labelPodCompleted
time="2023-05-22T14:10:10.961Z" level=info msg="cleaning up pod" action=labelPodCompleted key=ci-workflows/example-steps-simple-mutex4b72g-sleep-565248893/labelPodCompleted
time="2023-05-22T14:10:20.985Z" level=info msg="Processing workflow" namespace=ci-workflows workflow=example-steps-simple-mutex4b72g
time="2023-05-22T14:10:20.985Z" level=info msg="Task-result reconciliation" namespace=ci-workflows numObjs=4 workflow=example-steps-simple-mutex4b72g
time="2023-05-22T14:10:20.985Z" level=info msg="node changed" namespace=ci-workflows new.message= new.phase=Succeeded new.progress=0/1 nodeID=example-steps-simple-mutex4b72g-3623933960 old.message= old.phase=Running old.progress=0/1 workflow=example-steps-simple-mutex4b72g
time="2023-05-22T14:10:20.986Z" level=info msg="Workflow pod is missing" namespace=ci-workflows nodeName="example-steps-simple-mutex4b72g[0].job-1.hooks.success" nodePhase=Pending recentlyStarted=false workflow=example-steps-simple-mutex4b72g
time="2023-05-22T14:10:20.986Z" level=info msg="Running hooks" hookName=success lifeCycleHook=success namespace=ci-workflows node="example-steps-simple-mutex4b72g[0].job-1.hooks.success" workflow=example-steps-simple-mutex4b72g
time="2023-05-22T14:10:20.987Z" level=info msg="Workflow step group node example-steps-simple-mutex4b72g-681339196 not yet completed" namespace=ci-workflows workflow=example-steps-simple-mutex4b72g
time="2023-05-22T14:10:20.987Z" level=info msg="TaskSet Reconciliation" namespace=ci-workflows workflow=example-steps-simple-mutex4b72g
time="2023-05-22T14:10:20.987Z" level=info msg=reconcileAgentPod namespace=ci-workflows workflow=example-steps-simple-mutex4b72g
time="2023-05-22T14:10:21.000Z" level=info msg="Workflow update successful" namespace=ci-workflows phase=Running resourceVersion=365437539 workflow=example-steps-simple-mutex4b72g
time="2023-05-22T14:10:21.006Z" level=info msg="cleaning up pod" action=labelPodCompleted key=ci-workflows/example-steps-simple-mutex4b72g-hook-3623933960/labelPodCompleted
time="2023-05-22T14:10:30.892Z" level=info msg="cleaning up pod" action=killContainers key=ci-workflows/example-steps-simple-mutex4b72g-sleep-565248893/killContainers
time="2023-05-22T14:10:31.001Z" level=info msg="Processing workflow" namespace=ci-workflows workflow=example-steps-simple-mutex4b72g
time="2023-05-22T14:10:31.002Z" level=info msg="Task-result reconciliation" namespace=ci-workflows numObjs=4 workflow=example-steps-simple-mutex4b72g
time="2023-05-22T14:10:31.002Z" level=info msg="Workflow pod is missing" namespace=ci-workflows nodeName="example-steps-simple-mutex4b72g[0].job-1.hooks.success" nodePhase=Pending recentlyStarted=false workflow=example-steps-simple-mutex4b72g
time="2023-05-22T14:10:31.003Z" level=info msg="Running hooks" hookName=success lifeCycleHook=success namespace=ci-workflows node="example-steps-simple-mutex4b72g[0].job-1.hooks.success" workflow=example-steps-simple-mutex4b72g
time="2023-05-22T14:10:31.004Z" level=info msg="Workflow step group node example-steps-simple-mutex4b72g-681339196 not yet completed" namespace=ci-workflows workflow=example-steps-simple-mutex4b72g
time="2023-05-22T14:10:31.004Z" level=info msg="TaskSet Reconciliation" namespace=ci-workflows workflow=example-steps-simple-mutex4b72g
time="2023-05-22T14:10:31.004Z" level=info msg=reconcileAgentPod namespace=ci-workflows workflow=example-steps-simple-mutex4b72g
time="2023-05-22T14:10:31.015Z" level=info msg="Workflow update successful" namespace=ci-workflows phase=Running resourceVersion=365437539 workflow=example-steps-simple-mutex4b72g
time="2023-05-22T14:10:40.946Z" level=info msg="cleaning up pod" action=killContainers key=ci-workflows/example-steps-simple-mutex4b72g-hook-3623933960/killContainers
```

### Logs from in your workflow's wait container

```text
example-steps-simple-mutex4b72g-sleep-531693655: time="2023-05-22T14:10:00 UTC" level=info msg="Successfully saved file: /tmp/argo/outputs/logs/main.log"
example-steps-simple-mutex4b72g-sleep-548471274: time="2023-05-22T14:10:00 UTC" level=info msg="Successfully saved file: /tmp/argo/outputs/logs/main.log"
example-steps-simple-mutex4b72g-sleep-531693655: time="2023-05-22T14:10:00 UTC" level=info msg="Create workflowtaskresults 201"
example-steps-simple-mutex4b72g-sleep-548471274: time="2023-05-22T14:10:00 UTC" level=info msg="Create workflowtaskresults 201"
example-steps-simple-mutex4b72g-hook-3623933960: time="2023-05-22T14:10:02 UTC" level=info msg="Starting Workflow Executor" version=v3.4.7
example-steps-simple-mutex4b72g-hook-3623933960: time="2023-05-22T14:10:02 UTC" level=info msg="Using executor retry strategy" Duration=1s Factor=1.6 Jitter=0.5 Steps=5
example-steps-simple-mutex4b72g-hook-3623933960: time="2023-05-22T14:10:02 UTC" level=info msg="Executor initialized" deadline="2023-05-22 15:09:50 +0000 UTC" includeScriptOutput=
false namespace=ci-workflows podName=example-steps-simple-mutex4b72g-hook-3623933960 template="{\"name\":\"hook\",\"inputs\":{\"parameters\":[{\"name\":\"sleep_duration\",\"value\
":\"5\"},{\"name\":\"context\",\"value\":\"job-1\"}]},\"outputs\":{},\"metadata\":{},\"script\":{\"name\":\"\",\"image\":\"alpine:latest\",\"command\":[\"/bin/sh\"],\"resources\":
{},\"source\":\"echo \\\"Fake API call - sleeping for 5, context is job-1\\\"\\nsleep 5\\n\"},\"archiveLocation\":{\"archiveLogs\":true,\"s3\":{\"endpoint\":\"s3.amazonaws.com\",\
"bucket\":\"argo-workflows-artifacts-infrastructure-production\",\"region\":\"us-east-1\",\"insecure\":false,\"useSDKCreds\":true,\"key\":\"example-steps-simple-mutex4b72g/example
-steps-simple-mutex4b72g-hook-3623933960\"}},\"activeDeadlineSeconds\":1800,\"synchronization\":{\"mutex\":{\"name\":\"job-1\"}},\"timeout\":\"45m\"}" version="&Version{Version:v3
.4.7,BuildDate:2023-04-11T16:19:29Z,GitCommit:f2292647c5a6be2f888447a1fef71445cc05b8fd,GitTag:v3.4.7,GitTreeState:clean,GoVersion:go1.19.8,Compiler:gc,Platform:linux/amd64,}"
example-steps-simple-mutex4b72g-hook-3623933960: time="2023-05-22T14:10:02 UTC" level=info msg="Starting deadline monitor"
example-steps-simple-mutex4b72g-hook-3623933960: time="2023-05-22T14:10:08 UTC" level=info msg="Main container completed" error="<nil>"
example-steps-simple-mutex4b72g-hook-3623933960: time="2023-05-22T14:10:08 UTC" level=info msg="No Script output reference in workflow. Capturing script output ignored"
example-steps-simple-mutex4b72g-hook-3623933960: time="2023-05-22T14:10:08 UTC" level=info msg="No output parameters"
example-steps-simple-mutex4b72g-hook-3623933960: time="2023-05-22T14:10:08 UTC" level=info msg="No output artifacts"
example-steps-simple-mutex4b72g-hook-3623933960: time="2023-05-22T14:10:08 UTC" level=info msg="S3 Save path: /tmp/argo/outputs/logs/main.log, key: example-steps-simple-mutex4b72g
/example-steps-simple-mutex4b72g-hook-3623933960/main.log"
example-steps-simple-mutex4b72g-hook-3623933960: time="2023-05-22T14:10:08 UTC" level=info msg="Creating minio client using AWS SDK credentials"
example-steps-simple-mutex4b72g-hook-3623933960: time="2023-05-22T14:10:08 UTC" level=info msg="Saving file to s3" bucket=argo-workflows-artifacts-infrastructure-production endpoi
nt=s3.amazonaws.com key=example-steps-simple-mutex4b72g/example-steps-simple-mutex4b72g-hook-3623933960/main.log path=/tmp/argo/outputs/logs/main.log
example-steps-simple-mutex4b72g-hook-3623933960: time="2023-05-22T14:10:08 UTC" level=info msg="Save artifact" artifactName=main-logs duration=65.845848ms error="<nil>" key=exampl
e-steps-simple-mutex4b72g/example-steps-simple-mutex4b72g-hook-3623933960/main.log
example-steps-simple-mutex4b72g-hook-3623933960: time="2023-05-22T14:10:08 UTC" level=info msg="not deleting local artifact" localArtPath=/tmp/argo/outputs/logs/main.log
example-steps-simple-mutex4b72g-hook-3623933960: time="2023-05-22T14:10:08 UTC" level=info msg="Successfully saved file: /tmp/argo/outputs/logs/main.log"
example-steps-simple-mutex4b72g-hook-3623933960: time="2023-05-22T14:10:08 UTC" level=info msg="Create workflowtaskresults 201"
example-steps-simple-mutex4b72g-hook-3623933960: time="2023-05-22T14:10:08 UTC" level=info msg="Alloc=8189 TotalAlloc=19505 Sys=40813 NumGC=6 Goroutines=10"
```

**Comment by mnarozny** [comment]
Reproduced with version 3.4.8

**Comment by sarabala1979** [comment]
This issue will get resolved with #11113
